### DISCUSSION TITLE
How to extract texts between two coordinates in a page?

### DISCUSSION THREAD

**Discussion by StephenZKCurry** [pymupdf/PyMuPDF#3959] [Looking for help]
I want to extract texts between two coordinates on a page use the PDF's underlying flow of characters as a guide for ordering and segmenting the words, rather than presorting the characters by x/y position, mimics dragging a cursor highlights text in a PDF, How can I do that?

**Comment by JorjMcKie** [comment]
You can supply an arbitrary rectangle ("clip") inside which your desired text lives. If you only have top and bottom values, make a rectangle `clip = pymupdf.Rect(0, top, page.rect.width, bottom)`.
Then execute `text = page.get_text(sort=True, clip=clip)`.
This will (pymupdf v1.24.11+) extract the text in reading order.

**Comment by patsnap-liujin** [comment]
>  `clip = pymupdf.Rect(0, top, page.rect.width, bottom)`
`text = page.get_text(sort=True, clip=clip)`. 
This will (pymupdf v1.24.11+) extract the text in reading order.


Thanks for the answer!  @JorjMcKie I have some further questions😄. The `page.get_text` get the text ***intersecting*** the rect, but is there some encapsulated method or function to get the text ***fully contained*** in the rect ? I have read [textbox-extract-1.py](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/textbox-extraction/textbox-extract-1.py) and code an example below for my useage, and I'm looking for some practice more efficient and standard. Can you give some suggestions?
```python
# page = doc[page_num]
# my_rect = fitz.Rect(bbox)
current_textpage = page.get_textpage()
words = page.get_text("words",clip=my_rect, textpage=current_textpage) # here is the point: when using param `textpage` to reduce execution time significantly, the param clip is ignored and the result `words` contains whole page but I want the content only in my_rect.
fullly_containded_words = [w for w in words if fitz.Rect(w[:4]) in my_rect] # if I have many rects, the loop will  execute many times
fullly_containded_text = make_text(fullly_containded_words) # make_text is a function from textbox-extract-1.py
```

**Reply by patsnap-liujin** [reply]
After read the source code about `Page.get_textbbox`, I find the function `JM_rects_overlap` is the judgment of overlap. I change it from 
```python
def JM_rects_overlap(a, b):
    if (0
            or a.x0 >= b.x1
            or a.y0 >= b.y1
            or a.x1 <= b.x0
            or a.y1 <= b.y0
            ):
        return 0
    return 1
```
to
```python
def JM_rects_overlap(a, b):
    if (a.x0 < (b.x1+b.x0)/2 < a.x1
        and a.y0 < (b.y1+b.y0)/2 < a.y1):
        return 1
    return 0
```
and it works as I want. The change is judge from center point ,not edge point.
